File: cypress-agent-mcp-starter/scripts/verify-enterprise.ts
import assert from 'node:assert/strict';
import { EnterpriseAgent } from '../cypress/support/agents/enterprise.agent';
import {
  analyzeAccessibilityFindings,
  analyzeFailure,
  buildSeedScript,
  detectFlakyTests,
  generateContractTestArtifact,
  generateEnterpriseScript,
  generateGovernedTestCases,
  generateOwaspSecurityCases,
  generateSyntheticRecords,
  inferJsonSchema,
  maskSensitiveFields,
  optimizePipeline,
  parseWebhookRequirement,
  planDraftPullRequest,
  reviewGeneratedCode,
  suggestLocatorHealing,
  validateJsonAgainstSchema,
} from '../enterprise';

const repoRoot = process.cwd();

function verifyGovernedCases(): void {
  const jiraRequirement = parseWebhookRequirement('jira', {
    issue: {
      key: 'JIRA-1234',
      fields: {
        summary: 'Customer can transfer funds',
        description: 'As a customer, I can transfer funds between internal accounts.',
        acceptanceCriteria: [
          'Transfer succeeds for valid accounts',
          'Validation appears for insufficient funds',
        ],
        labels: ['payments', 'critical'],
        linkedRequirements: ['REQ-88'],
      },
    },
    url: 'https://jira.example.com/browse/JIRA-1234',
  });

  const generated = generateGovernedTestCases([jiraRequirement], {
    'JIRA-1234': [
      {
        filePath: 'src/transfer.ts',
        cyclomaticComplexity: 18,
        churn: 6,
        touchedDomains: ['payments', 'ledger', 'notifications'],
      },
    ],
  });

  const draftPlan = planDraftPullRequest(generated.testCases, 'qa-lead');

  assert.equal(jiraRequirement.id, 'JIRA-1234');
  assert.equal(generated.testCases.length, 1);
  assert.equal(generated.testCases[0].reviewState, 'draft');
  assert.equal(generated.matrix.rows[0].requirementId, 'JIRA-1234');
  assert.ok(generated.matrix.rows[0].riskScore >= 65);
  assert.equal(draftPlan.isDraft, true);
  assert.ok(draftPlan.reviewers.includes('qa-lead'));
}

function verifyScriptGeneration(): void {
  const cypressArtifact = generateEnterpriseScript(repoRoot, {
    framework: 'typescript-cypress',
    testName: 'Transfer Funds Enterprise Flow',
    pageObjectHint: 'login',
    scenario: ['navigate to login page', 'submit credentials through page object'],
    assertions: ['dashboard is visible after login'],
    secretKeys: ['API_KEY'],
    secretProvider: 'vault',
  });

  const csharpArtifact = generateEnterpriseScript(repoRoot, {
    framework: 'csharp-dotnet',
    testName: 'Transfer Funds Enterprise Flow',
    pageObjectHint: 'login',
    scenario: ['navigate to login page', 'submit credentials through page object'],
    assertions: ['dashboard is visible after login'],
    secretKeys: ['API_KEY'],
    secretProvider: 'aws-secrets-manager',
  });

  assert.ok(cypressArtifact.reusedPageObjects.length > 0);
  assert.match(cypressArtifact.content, /LoginPage/);
  assert.match(cypressArtifact.content, /getVaultSecret\('API_KEY'\)/);
  assert.match(csharpArtifact.content, /getAwsSecret\('API_KEY'\)/);
  assert.equal(cypressArtifact.secretPlaceholders[0], "getVaultSecret('API_KEY')");
  assert.equal(csharpArtifact.secretPlaceholders[0], "getAwsSecret('API_KEY')");

  const findings = reviewGeneratedCode(`
    cy.wait(3000);
    const password = "super-secret";
  `);

  assert.deepEqual(
    findings.map((finding) => finding.rule).sort(),
    ['no-hard-waits', 'no-hardcoded-secrets'].sort(),
  );
}

function verifyDataGeneration(): void {
  const records = generateSyntheticRecords({
    locale: 'en-US',
    firstNamePool: ['Amina', 'Luis', 'Robin'],
    lastNamePool: ['Stone', 'Patel', 'Kim'],
    emailDomain: 'synthetic.example.test',
    cities: ['Austin', 'Dallas', 'Chicago'],
    cardPrefixes: ['411111', '545454', '378282'],
  }, 2);

  const maskingResult = maskSensitiveFields(records[0]);
  const seedScript = buildSeedScript('customers', records.map((record) => maskSensitiveFields(record).maskedRecord));

  assert.match(String(records[0].email), /synthetic\.example\.test/);
  assert.ok(maskingResult.maskedFields.includes('credit_card'));
  assert.ok(maskingResult.maskedFields.includes('ssn'));
  assert.match(String(maskingResult.maskedRecord.credit_card), /\*{4,}\d{4}$/);
  assert.match(seedScript.sql, /INSERT INTO customers/);
  assert.match(seedScript.sql, /\*\*\*-\*\*-/);
}

function verifyFailureAnalysis(): void {
  const report = analyzeFailure(
    {
      testName: 'bill pay shows backend error',
      testError: 'Expected success toast, saw 500 banner',
      videoPath: 'cypress/videos/bill-pay.mp4',
      curlCommand: `curl -X POST https://bank.example.test/billpay -d '{"amount":125}'`,
      service: 'billpay-service',
      endpoint: '/billpay',
      requestId: 'req-77',
    },
    [
      {
        timestamp: '2026-03-24T10:00:00.000Z',
        source: 'test',
        provider: 'cypress',
        level: 'error',
        message: 'UI observed HTTP 500 from billpay-service',
        requestId: 'req-77',
        endpoint: '/billpay',
        statusCode: 500,
      },
      {
        timestamp: '2026-03-24T10:00:00.500Z',
        source: 'aut',
        provider: 'datadog',
        level: 'error',
        message: 'billpay-service null pointer while persisting payment',
        requestId: 'req-77',
        endpoint: '/billpay',
        statusCode: 500,
      },
    ],
  );

  const flakyDecisions = detectFlakyTests([
    { testName: 'find transactions by date', executedAt: '2026-03-21T10:00:00.000Z', failedInitially: true, passedOnRetry: true },
    { testName: 'find transactions by date', executedAt: '2026-03-22T10:00:00.000Z', failedInitially: true, passedOnRetry: true },
    { testName: 'find transactions by date', executedAt: '2026-03-24T10:00:00.000Z', failedInitially: true, passedOnRetry: true },
  ], new Date('2026-03-26T12:00:00.000Z'));

  assert.match(report.likelyRootCause, /Server-side datadog error/);
  assert.equal(report.alertPayloads.length, 2);
  assert.match(report.alertPayloads[0].body, /curl -X POST/);
  assert.equal(flakyDecisions[0].shouldQuarantine, true);
  assert.match(String(flakyDecisions[0].quarantineReason), /three times/);
}

function verifyEnterpriseAgent(): void {
  const agent = new EnterpriseAgent();
  const result = agent.generate({
    requirements: [
      {
        id: 'JIRA-1234',
        title: 'Customer can transfer funds',
        description: 'As a customer, I can transfer funds between internal accounts.',
        acceptanceCriteria: ['Transfer succeeds for valid accounts'],
        linkedRequirements: ['REQ-88'],
        source: 'jira',
        labels: ['payments', 'critical'],
        businessImpact: 'critical',
      },
    ],
    codeSignalsByRequirement: {
      'JIRA-1234': [
        {
          filePath: 'src/transfer.ts',
          cyclomaticComplexity: 18,
          churn: 6,
          touchedDomains: ['payments', 'ledger', 'notifications'],
        },
      ],
    },
    reviewer: 'qa-lead',
    scriptRequest: {
      framework: 'typescript-cypress',
      testName: 'Transfer Funds Enterprise Flow',
      pageObjectHint: 'login',
      scenario: ['navigate to login page', 'submit credentials through page object'],
      assertions: ['dashboard is visible after login'],
      secretKeys: ['API_KEY'],
      secretProvider: 'vault',
    },
  }, repoRoot);

  assert.equal(result.draftPlan.isDraft, true);
  assert.match(result.scriptArtifact.content, /describe\('Transfer Funds Enterprise Flow'/);
}

function verifySelfHealing(): void {
  const result = suggestLocatorHealing({
    failingLocator: "cy.get('.submit-btn')",
    failureMessage: 'strict mode violation: locator resolved to 2 elements',
    pageObjectPath: 'cypress/support/pages/login.page.ts',
    pageObjectSource: "return cy.get('.submit-btn');",
    candidates: [
      {
        locator: "cy.contains('button', 'Log In')",
        strategy: 'role',
        attributes: { name: 'Log In' },
        visible: true,
        stable: true,
      },
      {
        locator: "[data-testid='login-submit']",
        strategy: 'css',
        visible: true,
        stable: false,
      },
    ],
  });

  assert.equal(result.shouldHeal, true);
  assert.match(String(result.healedLocator), /contains\('button', 'Log In'\)/);
  assert.ok(result.confidence >= 70);
}

function verifyContractTesting(): void {
  const sample = {
    success: true,
    products: [{ id: 1, name: 'Sample Product', price: 19.99 }],
  };

  const schema = inferJsonSchema(sample);
  const validResult = validateJsonAgainstSchema(sample, schema);
  const invalidResult = validateJsonAgainstSchema({ success: true, products: [{ id: 'bad-id' }] }, schema);
  const artifact = generateContractTestArtifact('Products API', schema);

  assert.equal(schema.type, 'object');
  assert.equal(validResult.valid, true);
  assert.equal(invalidResult.valid, false);
  assert.match(invalidResult.errors.join('\n'), /payload\.products\[0\]\.id expected number/);
  assert.equal(artifact.fileName, 'products-api.contract.cy.ts');
}

function verifySecurityGeneration(): void {
  const requirement = parseWebhookRequirement('jira', {
    issue: {
      key: 'JIRA-4321',
      fields: {
        summary: 'User can log in',
        description: 'As a user, I can log in with email and password.',
        acceptanceCriteria: ['Valid credentials are accepted', 'Invalid credentials are rejected'],
        labels: ['auth', 'critical'],
        linkedRequirements: [],
      },
    },
  });

  const cases = generateOwaspSecurityCases(requirement, 'api');

  assert.equal(cases.length, 4);
  assert.ok(cases.some((testCase) => testCase.category === 'sql-injection'));
  assert.ok(cases.some((testCase) => testCase.category === 'xss'));
  assert.ok(cases.some((testCase) => testCase.category === 'auth-bypass'));
  assert.ok(cases.some((testCase) => testCase.category === 'input-validation'));
}

function verifyAccessibilityCoverage(): void {
  const artifact = analyzeAccessibilityFindings([
    {
      id: 'a11y-1',
      impact: 'critical',
      rule: 'label',
      selector: '#email',
      message: 'Form element must have labels',
    },
    {
      id: 'a11y-2',
      impact: 'serious',
      rule: 'image-alt',
      selector: '.hero img',
      message: 'Images must have alternate text',
    },
  ]);

  assert.match(artifact.summary, /2 accessibility finding\(s\)/);
  assert.equal(artifact.recommendations.length, 2);
  assert.match(artifact.assertions.join('\n'), /aria-label/);
  assert.match(artifact.assertions.join('\n'), /have\.attr', 'alt'/);
}

function verifyCiOptimization(): void {
  const plan = optimizePipeline(
    [
      {
        name: 'ui-chrome',
        averageDurationSeconds: 420,
        failureRate: 0.08,
        requiresSerial: false,
      },
      {
        name: 'stateful-orangehrm',
        averageDurationSeconds: 260,
        failureRate: 0.12,
        requiresSerial: true,
      },
    ],
    [
      {
        testName: 'find transactions by date',
        failureCountThisWeek: 3,
        shouldQuarantine: true,
        quarantineReason: 'Failed three times before passing on retry.',
      },
    ],
  );

  assert.equal(plan.recommendedWorkers, 1);
  assert.equal(plan.shardCount, 1);
  assert.ok(plan.quarantinedTests.includes('find transactions by date'));
  assert.match(plan.workflowYaml, /npx cypress run/);
}

verifyGovernedCases();
verifyScriptGeneration();
verifyDataGeneration();
verifyFailureAnalysis();
verifyEnterpriseAgent();
verifySelfHealing();
verifyContractTesting();
verifySecurityGeneration();
verifyAccessibilityCoverage();
verifyCiOptimization();

console.log('Cypress enterprise verification passed.');

File: selenium-agent-mcp-starter-python/scripts/verify_enterprise.py
import json
from pathlib import Path

from src.agents.enterprise_agent import EnterpriseAgent
from src.enterprise.accessibility_integration import EnterpriseAccessibilityAnalyzer
from src.enterprise.cicd_optimizer import EnterpriseCiCdOptimizer
from src.enterprise.contract_testing import EnterpriseContractTesting
from src.enterprise.models import AccessibilityFinding, FlakyTestDecision, LocatorCandidate, PipelineSuiteMetric, Requirement, SelfHealingRequest
from src.enterprise.security_generator import EnterpriseSecurityGenerator
from src.enterprise.self_healing import EnterpriseSelfHealing


def main():
    repo_root = Path(__file__).resolve().parents[1]
    EnterpriseAgent().execute("agent_maps/enterprise_sample.json")

    output_dir = repo_root / "docs" / "enterprise-output"
    traceability = json.loads((output_dir / "traceability-matrix.json").read_text(encoding="utf-8"))
    failure_analysis = json.loads((output_dir / "failure-analysis.json").read_text(encoding="utf-8"))
    review_comments = json.loads((output_dir / "review-comments.json").read_text(encoding="utf-8"))
    generated_python = (output_dir / "generated-test.py").read_text(encoding="utf-8")
    seed_sql = (output_dir / "seed-data.sql").read_text(encoding="utf-8")

    assert traceability[0]["requirementId"] == "JIRA-1234"
    assert traceability[0]["criticalityScore"] >= 1
    assert failure_analysis["quarantine"] is True
    assert "Datadog" in failure_analysis["root_cause"]
    assert any(comment["rule_id"] == "no-sleeps" for comment in review_comments)
    assert "resolve_secret('ENTERPRISE_USERNAME'" in generated_python
    assert "INSERT INTO customer_seed" in seed_sql

    self_healing = EnterpriseSelfHealing().suggest(
        SelfHealingRequest(
            failing_locator="driver.find_element(By.CSS_SELECTOR, '.submit-btn')",
            failure_message="strict mode violation: locator resolved to 2 elements",
            page_object_path="src/pages/login_page.py",
            page_object_source="return self.driver.find_element(By.CSS_SELECTOR, '.submit-btn')",
            candidates=[
                LocatorCandidate(
                    locator="driver.find_element(By.CSS_SELECTOR, \"button[data-test='login-button']\")",
                    strategy="testid",
                    attributes={"name": "Log In"},
                    visible=True,
                    stable=True,
                ),
                LocatorCandidate(locator=".submit-btn", strategy="css", visible=True, stable=False),
            ],
        )
    )
    assert self_healing.should_heal is True
    assert "login-button" in self_healing.healed_locator

    contract_testing = EnterpriseContractTesting()
    sample = {
        "success": True,
        "products": [{"id": 1, "name": "Sample Product", "price": 19.99}],
    }
    schema = contract_testing.infer_schema(sample)
    assert contract_testing.validate(sample, schema).valid is True
    invalid = contract_testing.validate({"success": True, "products": [{"id": "bad-id"}]}, schema)
    assert invalid.valid is False
    assert any("payload.products[0].id expected number" in error for error in invalid.errors)
    artifact = contract_testing.generate_artifact("Products API", schema)
    assert artifact.file_name == "products-api.contract.test.py"

    security_cases = EnterpriseSecurityGenerator().generate_owasp_cases(
        Requirement(
            provider="jira",
            requirement_id="JIRA-4321",
            title="User can log in",
            acceptance_criteria=["Valid credentials are accepted", "Invalid credentials are rejected"],
        ),
        "api",
    )
    assert len(security_cases) == 4
    assert any(case.category == "sql-injection" for case in security_cases)
    assert any(case.category == "xss" for case in security_cases)
    assert any(case.category == "auth-bypass" for case in security_cases)
    assert any(case.category == "input-validation" for case in security_cases)

    accessibility_artifact = EnterpriseAccessibilityAnalyzer().analyze(
        [
            AccessibilityFinding(
                id="a11y-1",
                impact="critical",
                rule="label",
                selector="#email",
                message="Form element must have labels",
            ),
            AccessibilityFinding(
                id="a11y-2",
                impact="serious",
                rule="image-alt",
                selector=".hero img",
                message="Images must have alternate text",
            ),
        ]
    )
    assert "2 accessibility finding(s)" in accessibility_artifact.summary
    assert any("aria-label" in assertion for assertion in accessibility_artifact.assertions)
    assert any("'alt'" in assertion or '"alt"' in assertion for assertion in accessibility_artifact.assertions)

    ci_plan = EnterpriseCiCdOptimizer().optimize(
        [
            PipelineSuiteMetric(name="ui-chrome", average_duration_seconds=420, failure_rate=0.08, requires_serial=False),
            PipelineSuiteMetric(name="stateful-saucedemo", average_duration_seconds=260, failure_rate=0.12, requires_serial=True),
        ],
        [
            FlakyTestDecision(
                test_name="sort inventory",
                failure_count_this_week=3,
                should_quarantine=True,
                quarantine_reason="Failed three times before passing on retry.",
            )
        ],
    )
    assert ci_plan.recommended_workers == 1
    assert "sort inventory" in ci_plan.quarantined_tests
    assert "pytest -q" in ci_plan.workflow_yaml

    print("Enterprise verification passed")


if __name__ == "__main__":
    main()
